In [77]:
import pandas as pd
import plotly.graph_objects as go
import numpy as np

JI = pd.read_csv('JPY_INR Historical Data.csv', index_col=0)
UI = pd.read_csv('USD_INR Historical Data.csv', index_col=0) 
UJ = pd.read_csv('USD_JPY Historical Data.csv', index_col=0) 

In [78]:
fig = go.Figure()

fig.add_scatter(x=JI.index, y=JI['Price'], name='JPY/INR')
fig.add_scatter(x=UI.index, y=UI['Price']/UJ['Price'], name='JPY/INR')

In [79]:
fig = go.Figure()

fig.add_scatter(x=JI.index, y= abs(JI['Price'] - UI['Price']/UJ['Price']), name='JPY/INR')

Theres an obvious arb lmao

In [80]:
JI = JI[JI.index == "May 24, 2022"]['Price'][0]
UI = UI[UI.index == "May 24, 2022"]['Price'][0]
UJ = UJ[UJ.index == "May 24, 2022"]['Price'][0]

In [142]:
from math import log

v = [[1, UI, UJ], [1/UI, 1, 1/JI], [1/UJ, JI, 1]]
v = [[log(e) for e in r] for r in v]


M = np.matrix(v)

M

matrix([[ 0.        ,  4.34976167,  4.84261104],
        [-4.34976167,  0.        ,  0.49200387],
        [-4.84261104, -0.49200387,  0.        ]])

In [144]:
from scipy.optimize import linear_sum_assignment
row_ind, col_ind = linear_sum_assignment(M, maximize=True)

print(M[row_ind, col_ind].sum())

print(list(zip(row_ind, col_ind)))

0.0008454983911163061
[(0, 2), (1, 0), (2, 1)]


USD -> JPY ( 4.84) 
JPY -> INR (-4.34)
INR -> USD (-0.49)
            +0.0008

Assuming we want to trade all of them

Work on the Forex Tree Trading, start with making a class etc

In [ ]:
def invest(root, path):
    
    for child in root.children:
        invest(child  root)
